teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -171,12 +195,399 @@ In addition to grouping related resources, Tekton Results provides long term sto
+spec:
+    ...
+status:
+    completionTime: "2020-05-04T02:19:14Z"
+    conditions:
+      - lastTransitionTime: "2020-05-04T02:19:14Z"
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    taskRuns:
+      triggers-release-nightly-build:
+        pipelineTaskName: build
+        status:
+          completionTime: "2020-05-04T02:10:49Z"
+          conditions:
+            - lastTransitionTime: "2020-05-04T02:10:49Z"
+              message: All Steps have completed executing
+              reason: Succeeded
+              status: "True"
+              type: Succeeded
+          podName: triggers-release-nightly-build-pod
+          resourcesResult:
+            - key: commit
+              resourceRef:
+                name: git-source-triggers
+              value: 9ab5a1234166a89db352afa28f499d596ebb48db
+          startTime: "2020-05-04T02:05:07Z"
+          steps:
+            - container: step-build
+              imageID: docker-pullable://golang@sha256:a90f267133
+              name: build
+              terminated:
+                containerID: docker://6b6471f501f59dbb
+                exitCode: 0
+                finishedAt: "2020-05-04T02:10:45Z"
+                reason: Completed
+                startedAt: "2020-05-04T02:06:24Z"
+```
+
+Taking the above example, this will be the new minimal `PipelineRun` status:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  ...
+spec:
+    ...
+status:
+    completionTime: "2020-05-04T02:19:14Z"
+    conditions:
+      - lastTransitionTime: "2020-05-04T02:19:14Z"
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    taskRunsStatuses:
+      - taskRunName: triggers-release-nightly-build
+        conditions:
+          - lastTransitionTime: "2020-05-04T02:10:49Z"
+            message: All Steps have completed executing
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+```
+
+### Beta API
+
+Because the `PipelineRun` status is part of the [Pipelines API][api-definition],
+[removing the full embedded statuses](#2-deprecate-and-remove-full-embedded-status)
+is backwards incompatible. 
+
+To support migration as required by our [API compatibility policy][deprecations]
+we will add a behavior flag - `embedded-status` - used to configure whether the
+`PipelineRuns` should contain the full embedded status or the minimal embedded
+status of its `TaskRuns` and `Runs`.
+
+Following our [policy][behavior-flags] on updating behavior flags:
+1. The `embedded-status` flag will be `full` by default, users can set it to `minimal`. 
+2. After 9 months in v1beta1, the `embedded-status` flag will be changed to `minimal`
+  by default.
+3. As soon as the next release in v1beta1, `embedded-status` flag will be removed as
```

Comment:
Does this mean we have only one release when the `embedded-status` flag has it's default value changed ? 🤔

Reply:
This is saying that we can have as little as one release, but we can have more if we choose. I think this is allowable based on my reading of our compatibility policy on feature flags.

Reply:
I would rather have more time there aka the same amount of time we had between adding this flag and switching its default.
Just one release might force us (openshift-pipelines) to patches tekton version to allow a smooth transition for our users (as we usually do a release each 3/4 release of pipeline)

Reply:
giving it 9 more months means that we'd be taking 18 months to make the backwards incompatible change

while the [policy](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md#promoting-behavior-flags) says as soon as the next release, in reality it usually takes a longer after we confirm with users and contributors in working group meetings and mailing lists that they are ready for it to be removed -- added this detail in the migration plan

Reply:
@jerop so for me it's 2 backward compatibility changes though 😛 : 1 changes the default value, the other remove a flag 😛

Reply:
Interesting 🤔 

I see the backwards incompatible change as one thing: removing the full embedded status fields

The flag is just a means to make the change in a backwards compatible way initially and allow users to migrate to the minimal embedded statuses over the 9 months, I believe this is why [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md#promoting-behavior-flags) says:


> After 9 months, we change the default value:
> 1. We change the default value of `workspace-dir-rename` to true
> 2. In the release notes we warn that this is about to be changed - if people want to maintain the previous behavior
   they will need to explicitly set the flag to false
> 3. When the flag is set to false, a warning will be logged
>
> At this point we consider the backwards incompatible change to have been made, and we can decide when to finally remove the flag entirely (e.g. as soon as the next release).

Reply:
we could specify in this proposal that we'll provide at least 3 months after we remove the flag, what do you think?

Reply:
Our team at IBM is relying on the Tekton golang client to get all the status from pipelinerun. Is it possible to have a helper function in the golang client to retrieve all the `embedded-status` after the transition period? This way golang developers don't have to maintain their own way to get all the status and this function can be used in Tekton CLI as well.

Reply:
@vdemeester @Tomcli will this alternative migration plan be better for the migration instead of the current plan? 
https://github.com/jerop/community/blob/tep-0100-proposal/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md#beta-api---provide-both-full-and-minimal-statuses

Reply:
We are okay with the migration plan. We just have similar use cases as Tekton CLI where we need to know the pod name from taskruns to stream logs. We feel like many other developers might need to do the same migration, so it could be better if we can have some common functions to get all status in golang.

Reply:
sounds good, I've updated the TEP to include [Go libraries](https://github.com/jerop/community/blob/tep-0100-proposal/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md#go-libraries) and to [provide at least three more months before flag is removed](https://github.com/jerop/community/blob/tep-0100-proposal/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md#beta-api)